src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -119,21 +120,50 @@ class Backend(object):
     """
     name = "openssl"
 
+    # FIPS doesn't allow several algorithms or weak key sizes, but the
+    # algorithms are still present in OpenSSL. They just errors if you try
+    # to use them. To avoid that we return from methods like hash_supported.
+    _non_fips_aead = {b'chacha20-poly1305'}
+    _non_fips_ciphers = (
+        ARC4, Blowfish, CAST5, Camellia, ChaCha20, IDEA, SEED
+    )
+    _non_fips_hashes = (hashes.MD5, hashes.BLAKE2b, hashes.BLAKE2s)
+    _rsa_min_key_size = 2048
+    _rsa_min_public_exponent = 65537
+    _dsa_min_modulus = 1 << 2048
+    _dh_min_key_size = 2048
+    _dh_min_modulus = 1 << _dh_min_key_size
+
     def __init__(self):
         self._binding = binding.Binding()
         self._ffi = self._binding.ffi
         self._lib = self._binding.lib
+        self._fips_enabled = self._is_fips_enabled()
```

Comment:
This means if you try to enable FIPS later, won't something confusing happen?

Reply:
For now, yes. It's easier to require that it be in FIPS mode at startup (so we don't try to add the engine at all), so let's just make that a requirement.